text/0000-app-mixins.md

```diff
@@ -0,0 +1,241 @@
+
+## Rebasing an App
+
+Before a launch image is rebased, the platform must re-run the any stackpacks that were used to build the launch image against the new run-image. Then, the rebase operation can be performed as normal, while including the stackpack layers as part of the stack. This will be made possible by including the stackpack in the run-image, but because the stackpack detect phase is not run, the operation does not need access to the application source.
+
+## Example: Apt Buildpack
+
+A buildpack that installs a custom certificate would have a `buildpack.toml` like this:
+
+ ```toml
+[buildpack]
+id = "example/apt"
+privileged = true
+mixins = ["[^=]+"]
+
+[[stacks]]
+id = "io.buildpacks.stacks.bionic"
+```
+
+Its `bin/detect` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+exit 0
+```
+
+Its `bin/build` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+apt update -y
+
+for package in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
+  apt install $package
+done
+
+cat << EOF > launch.toml
+[[excludes]]
+paths = [ "/var/cache" ]
+cache = true
+EOF
+```
+
+## Example: CA Certificate Buildpack
+
+A buildpack that works on the `ubuntu-20` stack, which defines a `type=` prefix for mixins, and installs customer CA Certificates would have a `buildpack.toml` that looks like this:
+
+```toml
+[buildpack]
+id = "example/cacerts"
+privileged = true
+mixins = ["type=cacert"]
+
+[[stacks]]
+id = "ubuntu-20"
+```
+
+Its `bin/detect` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+exit 0
+```
+
+Its `bin/build` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+for cert_path in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
+  cert_file=$(basename $cert_path)
+  cp $cert_path /usr/share/ca-certificates/$cert_file
+  chmod 644 /usr/share/ca-certificates/$cert_file
+done
+
+update-ca-certificates
+```
+
+An application that is using this stackpack would the following `project.toml`:
+
+```toml
+[build]
+mixins = [ "type=cacert:mycerts/database.crt" ]
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Mixins are less flexible that a generic [root buildpack](https://github.com/buildpacks/rfcs/blob/app-image-ext-buildpacks/text/0000-app-image-extensions.md).
+
+# Alternatives
+[alternatives]: #alternatives
+
+- [Add root buildpack as interface for app image extension RFC](https://github.com/buildpacks/rfcs/pull/77)
+- [App Image Extensions (OS packages)](https://github.com/buildpacks/rfcs/pull/23)
+- [Root Buildpacks](https://github.com/buildpacks/rfcs/blob/root-buildpacks/text/0000-root-buildpacks.md): Allow application developers to use root buildpacks in `project.toml`. This would have significant performance implications, and creates a "foot gun" in qhich end users could build images they are not able to maintain. For this reason, we are holding off on a generic root buildpack feature.
```

Comment:
```suggestion
- [Root Buildpacks](https://github.com/buildpacks/rfcs/blob/root-buildpacks/text/0000-root-buildpacks.md): Allow application developers to use root buildpacks in `project.toml`. This would have significant performance implications, and creates a "foot gun" in which end users could build images they are not able to maintain. For this reason, we are holding off on a generic root buildpack feature.
```